nextcloud: add licensing information for plugins

RaitoBezarius opened the pull request:
###### Description of changes

Opened as a draft because I won't have time to finish it, cc @onny please pick it up.

There is no reason to not carry licenses and other information to any patched source.

This should improve our metadata in those situations.

Closes #238402.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
trivial-builders/applyPatches: carry `meta` information to the patched source

There is no reason to not carry licenses and other information to any patched source.

This should improve our metadata in those situations.

RaitoBezarius committed:
fetchNextcloudApp: meta propagation for licenses, etc.

This improves the metadata propgation for nc4nix-generated packages.

Adds:

- licenses (best effort given spdxId doesn't seem guaranteed here)
- homepage
- longDescription

RaitoBezarius commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -1,10 +1,20 @@
 }:
+let
+  # TODO: do something better
+  licenseMap = {
```
@alyssais could I have your opinion on approximating licenses this way? Is this bad or is there a way to make it OK?

mweinelt replied:
I don't suppose the upstream metadata could help clear this up?

RaitoBezarius replied:
I see https://github.com/nextcloud/bookmarks/blob/master/COPYING so I think this is a nc4nix improvement maybe? It should get SPDX IDs as much as possible.

mweinelt replied:
https://github.com/nextcloud/bookmarks/blob/master/package.json#L50

Ma27 replied:
cc @armijnhemel

RaitoBezarius committed:
nextcloud*Packages: add description, homepage, licenses from JSON

This ensures this data is replicated in the meta attribute.

Ma27 commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -878,13 +878,17 @@ rec {
-    };
+    }
+    # Carry `meta` information from the underlying `src` if present.
+    // (optionalAttrs (builtins.hasAttr "meta" src) src.meta)
```
What's wrong with `src?meta`?

Ma27 commented on pkgs/build-support/trivial-builders/default.nix:
```diff
@@ -878,13 +878,17 @@ rec {
+    }
+    # Carry `meta` information from the underlying `src` if present.
+    // (optionalAttrs (builtins.hasAttr "meta" src) src.meta)
+    // (removeAttrs args [ "src" "name" "patches" "postPatch" ]);
```
Quick note: I'm not sure how often this is actually used, but if this discussion re license mapping takes a little longer, we may want to consider pushing this part of the change to master already, I can imagine that there are more places where this would fix an issue with missing licenses.

RaitoBezarius commented:
3 weeks have passed with no reaction from @onny — @Ma27 are you interested into picking the pieces? Otherwise, I feel like we should discuss removal of plugins or a way out of the current situation.

Ma27 commented:
@RaitoBezarius I'll do my best, two issues though: first, time is slightly limited, so it may take a moment. Second, I'll need some input on the licensing topic because I don't feel qualified to decide if your suggested approach is correct or not.

Ma27 committed:
applyPatches: fix adding meta information

Ma27 committed:
nextcloud*Packages: expose proper license information

This change makes sure that each Nextcloud plugin installed provides a
`meta` section with proper license information.

Unfortunately, the metadata from the appstore is useless for this
purpose since it doesn't differentiate between e.g. AGPL 3.x and AGPL
3.x or any later version. In fact, this isn't consistent between their
software, e.g. `bookmarks` has `agpl3Plus` according to the files'
headers[1] whereas `twofactor_nextcloud_notification` is AGPL 3 only[2].

I don't think there's any trivial and reasonable way to retrieve this
information programatically, so I decided to change the format of
`nextcloud-apps.json`[3] to also contain the license in the form of the
license attribute we have in `lib/licenses.nix`, i.e. GNU AGPL 3 or
later is `agpl3Plus`.

I retrieved the information using the following approach:

* go to the source code of $app at the revision we currently have
  packaged
* check for a license identifier (does it specify the license only or
  the license "or any later version")?
  * first in `src/main.js` because most apps from Nextcloud itself used
    actual spdx identifiers in the frontend's source-code.
  * then in `lib/AppInfo/Application.php` which each Nextcloud app has.

License changes should be updated accordingly when updating the apps. As
with any other package in nixpkgs as well, this currently needs to be
done manually (or as part of the review process)[4].

Also, I decided to change the `name` of the `applyPatches` derivation
from `source-patched` to `nextcloud-app-${appName}-${appVersion}`. When
deploying a lot of apps (and probably displaying the diff using
`nix store diff-closures` on deploy), the current output isn't very
helpful. This is purely optional because I didn't want to break the
interface of `fetchNextcloudApp` again.

[1] https://github.com/nextcloud/bookmarks/blob/v13.1.0/lib/AppInfo/Application.php#L6
[2] https://github.com/nextcloud/twofactor_nextcloud_notification/blob/v3.7.0/lib/AppInfo/Application.php
[3] This isn't really well-defined since it's preprocessed with `jq(1)`
    before passing the apps to nc4nix.
[4] Though notable license changes (e.g. agpl -> gpl) would also pop up
    in the diff of <nextcloudversion>.json, so this is pretty easy to
    catch.

Ma27 committed:
nextcloud*Packages: update

Ma27 committed:
fetchNextcloudApp: remove backwards compat for old interface

The `name` & `version` attributes only existed in a previous form of
the API before it was switched over to using `fetchzip` &
`applyPatches`[1]. The attributes existed to be able to throw an
evaluation error with upgrade instructions when this was used. However,
this was before 22.11, so this shouldn't be in use anymore (and if
somebody is doing a migration from a very old NixOS, this is still
documented in the 22.11 release-notes[2]).

Anyways, this simplifies the code a little bit and also having both
`appName`/`appVersion` and `name`/`version` in there is quite confusing. But
still, I figured it's less confusing to not re-use attributes that were
deprecated in the past, hence the alternative naming.

[1] 3ca9b9a8ad1b9dee2ec40eecca557f0578786b93
[2] d41b381310f5dcde1434b23406532f54ebdeea22

Ma27 commented:
@RaitoBezarius long story short, since one Nextcloud app has `agpl3Only` whereas everything else seems to have `agpl3Plus`, I decided that it doesn't make much sense to do this programmatically given the incomplete metadata from Nextcloud. Instead, the correct license must be specified manually now in a way that when omitting this, the evaluation will fail.

Further implementation notes can be found in the commit messages.

also cc @mweinelt @bachp @dasJ

RaitoBezarius reviewed (commented):
LGTM.

frederictobiasc commented:
This change breaks my configuration. 
As described in the [wiki](https://nixos.wiki/wiki/Nextcloud#Apps) I add apps to nextcloud, via extraApps:

eg.:
```
    extraApps = {
      spreed = pkgs.fetchNextcloudApp rec {
        url = "https://github.com/nextcloud-releases/spreed/releases/download/v17.0.2/spreed-v17.0.2.tar.gz";
        sha256 = "sha256-TJuQYQQGjXcf4YIsP0mIOlJA8DB8z698FfWU9Os2bh0=";
      };
```
Due to breaking changes, I now need to also provide `license` as an argument for `fetchNextcloudApp.
Shouldn't breaking changes be documented somewhere?

RaitoBezarius commented:
We should definitely document it in the release notes, it's not obvious
because it's kind of private API for nixpkgs usage and there's no real API
stability on those.

Either case, let's document it, yes.

Le jeu. 31 août 2023 à 21:59, Frédéric Tobias Christ <
***@***.***> a écrit :

> This change breaks my configuration.
> As described in the wiki <https://nixos.wiki/wiki/Nextcloud#Apps> I add
> apps to nextcloud, via extraApps:
>
> eg.:
>
>     extraApps = {
>       spreed = pkgs.fetchNextcloudApp rec {
>         url = "https://github.com/nextcloud-releases/spreed/releases/download/v17.0.2/spreed-v17.0.2.tar.gz";
>         sha256 = "sha256-TJuQYQQGjXcf4YIsP0mIOlJA8DB8z698FfWU9Os2bh0=";
>       };
>
> Due to breaking changes, I now need to also provide license as an
> argument for `fetchNextcloudApp.
> Shouldn't breaking changes be documented somewhere?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/238407#issuecomment-1701597340>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACMZRA5LTFSQK3NHAR73Y3XYDNKTANCNFSM6AAAAAAZK3IMBI>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

Stunkymonkey commented:
@RaitoBezarius I have a question about the examples in:
https://github.com/NixOS/nixpkgs/blob/df9376911996e3deab3db3e01b9cfe7399ade856/nixos/modules/services/web-apps/nextcloud.nix#L126-L141

I feel like this was forgotten? right?
